### DISCUSSION TITLE
GraphQL Custom Argument

### DISCUSSION THREAD

**Discussion by mortscode** [craftcms/cms#9851] [Q&A] [answered]
## TL;DR
I built a plugin that gives users a space to leave feedback (`name`, `email`, `comment`, `rating`, etcetera). The FeedbackElement is a custom Element (`craft\base\Element`) and I'm using CraftPro's GraphQL setup to provide data to the frontend for a "Load More" experience. I'm wondering if there's a way to create a custom GraphQL Argument to filter out Feedback items where the user left the `comment` value empty. Something like:
```
feedback (hasComment: true) {
    name
    comment
}
```

## What's working
I'm using the approach defined in the Extending GraphQL docs (https://craftcms.com/docs/3.x/extend/graphql.html#arguments) to create some custom arguments. For instance, I'm using the following query successfully in my JS:
```
{
    feedback(
        feedbackType: "review",
        feedbackStatus: "approved",
        limit: 5,
    ) {
        name
        comment
        rating
        response
        feedbackType
        feedbackStatus
    }
}
```
In the above query, `feedbackType` and `feedbackStatus` are both custom arguments from my FeedbackElement. I'm extending the `StructureElementArguments` with the `getArguments()` method like so and all is working as expected:
```
public static function getArguments(): array
{
    return array_merge(parent::getArguments(), self::getContentArguments(), [
        [...]
        'feedbackStatus' => [
            'name' => 'feedbackStatus',
            'type' => Type::listOf(Type::string()),
            'description' => 'Narrows the query results based on the feedback status.'
        ],
        'feedbackType' => [
            'name' => 'feedbackType',
            'type' => Type::listOf(Type::string()),
            'description' => 'Narrows the query results based on the type of feedback.'
        ],
        [...]
    ]);
}
```

## My Question
If possible, I'd like to create a custom Argument with some conditional logic in it. I'm hoping to only return the Feedback items where a user has entered a value in the `comment` field. Something like this:
```
{
    feedback(
        feedbackType: "review",
        feedbackStatus: "approved",
        hasComment: true, <--- new bool
        limit: 5,
    ) {
        name
        comment
        rating
        response
        feedbackType
        feedbackStatus
    }
}
```
The datapoint `hasComment` does not exist on my Element. My hope would be to return `false` if the `comment` value was an empty string (`""`). The above Arguments I've created for `feedbackType` and `feedbackStatus` simply use the value of those fields (no conditional logic) and I can't find an example of how to include a conditional operator in the creation of an Argument.

Is this possible? Is there a better way to approach this like maybe in my GQL query? Help.

**Comment by andris-sevcenko** [comment]
Depending on how you have set things up, but most likely you have a custom resolver set up since it's a custom element, so it's entirely up to you what you do with the arguments that are passed in.

For example, the entry resolver knows that any arguments are going to have matching methods on the element query, so it just [passes those values onwards](https://github.com/craftcms/cms/blob/develop/src/gql/resolvers/elements/Entry.php#L42-L44).

You likely have a custom element query class, so it's mostly a matter of implementing this on that class and then passing the value in the resolver to the element query.

**Reply by mortscode** [reply]
Thanks @andris-sevcenko

If I were to make a custom element query class, would that exist on my element's `src/elements/db/FeedbackElementQuery` file or is it a different file entirely? Any examples of how to create such a class?

By the way, here is my plugin: https://github.com/mortscode/feedback-plugin

**Reply by mortscode** [reply]
In the docs, it says that "conditions should be attached to a subquery": https://craftcms.com/docs/3.x/extend/element-types.html#this-query-vs-this-subquery

Could I just add a new subquery for `hasComment` in my `beforePrepare()` method on the `FeedbackElementQuery` class? https://github.com/mortscode/feedback-plugin/blob/73f062a5680b1cf1494ff2a4cbf9a2d93fa277f8/src/elements/db/FeedbackElementQuery.php#L89

Any idea what that might look like? This is not working: 
```
if ($this->hasComment) {
    $this->subQuery->andWhere(Db::parseParam(
        'feedback_record.comment',
        !empty($this->comment))
    );
}
```
The subquery examples in the docs are passing the value of the field in the db table. I know I could add a new table column for `hasComment` and create a migration to update the db columns, but it feels redundant to me. I could be wrong.

**Reply by mortscode** [reply]
Ok. I've set up a custom subquery for `hasComment` and now I can query on the frontend twig templates successfully with:
```
craft.feedback.getEntryFeedback(entry.id).hasComment(false).all()
```
I'm having trouble getting the same query to work successfully as an argument in GQL

**Comment by mortscode** [comment, marked as answer]
YOOOOO! Got this working like a dream.

Worked with @AugustMiller to figure out a solution to add the custom query (computed value) to my ElementQuery and then I was able to add the Argument to my plugin's GQL setup. Very stoked.

The ElementQuery approach:
https://gist.github.com/AugustMiller/7a0c5f87019f5d959b0d6001ba0be9b0

And then I added it to my `src/gql/interfaces` and `src/gql/arguments` files following the methods in the Extending GQL docs: 
- https://craftcms.com/docs/3.x/extend/graphql.html#interfaces
- https://craftcms.com/docs/3.x/extend/graphql.html#arguments

**Reply by andris-sevcenko** [reply]
Perfect! You guys make a great team!
